avr-hal-generic/src/spi.rs

```diff
@@ -0,0 +1,186 @@
+//! SPI Implementation
+
+pub use embedded_hal::spi;
+
+/// Oscillator Clock Frequency division options.  Controls both SPR and SPI2X register bits.
+pub enum SerialClockRate {
+    OscfOver2,
+    OscfOver4,
+    OscfOver8,
+    OscfOver16,
+    OscfOver32,
+    OscfOver64,
+    OscfOver128,
+}
+
+/// Order of data transmission, either MSB first or LSB first
+pub enum DataOrder {
+    MostSignificantFirst,
+    LeastSignificantFirst,
+}
+
+/// Settings to pass to Spi.
+///
+/// Easiest way to initialize is with
+/// `Settings::default()`.  Otherwise can be instantiated with alternate
+/// settings directly.
+pub struct Settings {
+    pub data_order: DataOrder,
+    pub clock: SerialClockRate,
+    pub mode: spi::Mode,
+}
```

Comment:
Please derive all traits for these structs/enums that make sense.  E.g. `Debug`, `Clone`, `Copy` (just for the enums), `PartialEq`, and `Eq`.